qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -14,233 +14,393 @@
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        if self._basis == 'XYX':
-            return self._circuit_xyx(theta,
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        raise QiskitError("OneQubitEulerDecomposer: invalid basis")
+    @property
+    def basis(self):
+        """The decomposition basis."""
+        return self._basis
+
+    @basis.setter
+    def basis(self, basis):
+        """Set the decomposition basis."""
+        basis_methods = {
+            'U3': (self._params_u3, self._circuit_u3),
+            'U1X': (self._params_u1x, self._circuit_u1x),
+            'RR': (self._params_zyz, self._circuit_rr),
+            'ZYZ': (self._params_zyz, self._circuit_zyz),
+            'ZXZ': (self._params_zxz, self._circuit_zxz),
+            'XYX': (self._params_xyx, self._circuit_xyx)
+        }
+        if basis not in basis_methods:
+            raise QiskitError("OneQubitEulerDecomposer: unsupported basis {}".format(basis))
+        self._basis = basis
+        self._params, self._circuit = basis_methods[self._basis]
+
+    def angles(self, unitary):
+        """Return the Euler angles for input array.
+
+        Args:
+            unitary (np.ndarray): 2x2 unitary matrix.
+
+        Returns:
+            tuple: (theta, phi, lambda).
+        """
+        theta, phi, lam, _ = self._params(unitary)
+        return theta, phi, lam
+
+    def angles_and_phase(self, unitary):
+        """Return the Euler angles and phase for input array.
+
+        Args:
+            unitary (np.ndarray): 2x2 unitary matrix.
+
+        Returns:
+            tuple: (theta, phi, lambda, phase).
+        """
+        return self._params(unitary)
+
+    def circuit(self, theta, phi, lam, simplify=True,
+                simplify_tolerance=DEFAULT_SIMPLIFY_TOLERANCE):
+        """Return the basis circuit for the input parameters.
+
+        Args:
+            theta (float): euler angle parameter
+            phi (float): euler angle parameter
+            lam (float): euler angle parameter
+            simplify (bool): simplify output circuit [Default: True]
+            simplify_tolerance (float): absolute tolerance for checking
+                                        angles zero [Default: 1e-12].
+
+        Returns:
+            QuantumCircuit: the basis circuits.
+        """
+        return self._circuit(theta, phi, lam,
+                             simplify=simplify,
+                             simplify_tolerance=simplify_tolerance)
 
     @staticmethod
-    def _angles_zyz(unitary_mat):
-        """Return euler angles for special unitary matrix in ZYZ basis.
+    def check_equiv(unitary, circuit, phase_equal=False, atol=DEFAULT_ATOL):
```

Comment:
this should be a private method, and actually it should be removed I think. We have extensive testing of this decomposer so there's no point slowing it down at the point of use.
I think a function like this could be useful in the tests: `test/python/quantum_info/test_synthesis.py`

Reply:
also doing this has the added bonus that `atol` can be removed (in which case you might choose to just keep `atol` instead of your current `simplify_tolerance`).

Reply:
Sounds good, can move this out